feat: implement a /neetcode command displaying the NeetCode code solution for the selected problem and programming language

cf841 opened the pull request:
## Changes
- Implemented the function /neetcode, which takes a programming language covered by neetcode, and a problem from leetcode, either by id, name or URL
- Get's the ID and title for the problem, from the leetcode graphql, and uses this to find the corresponding solution in the neetcode git repository. 
- generalized `ProblemSearchModal` so that different commands can call this one function and get the specific data they need from the graphql, by passing an embed function. 
- Added status checking to fetch_data
- Made `search_problem` more useful by returning title and id, rather than just title.
- Updated /help.

Kevin-Roman commented on cogs/neetcode.py:
```diff
@@ -0,0 +1,28 @@
+        self.bot = bot
+
+    @app_commands.command(name="neetcode")
+    async def neet(self, interaction: discord.Interaction) -> None:
```
call the function `neetcode`

Kevin-Roman commented on cogs/problems.py:
```diff
@@ -55,6 +55,7 @@ async def random_problem(
         :param difficulty: The desired difficulty level
         """
         embed = await random_question_embed(self.bot, difficulty.value)
+        
```
unnecessary change to this file

Kevin-Roman commented on ui/embeds/problems.py:
```diff
@@ -35,6 +36,14 @@ async def search_question_embed(bot: "DiscordBot", search_text: str) -> discord.
     embed = await question_embed(bot, question_title)
     return embed
 
+# Neetcode Search by id
+async def neetcode_search_embed(bot: "DiscordBot", search_text: str) -> discord.Embed:
+    question_title, question_id = await search_question(bot, search_text)
+
+    if not question_id:
+        return question_error_embed()
+    embed = await neetcode_embed(bot, question_id, question_title)
+    return embed
```
Move this to `embeds/neetcode.py`

Kevin-Roman commented on ui/modals/neetcode.py:
```diff
@@ -0,0 +1,28 @@
+from typing import TYPE_CHECKING
+
+import discord
+
+from ui.embeds.problems import neetcode_search_embed
+
+if TYPE_CHECKING:
+    # To prevent circular imports
+    from bot import DiscordBot
+
+
+class NeetcodeSearchModal(discord.ui.Modal, title="Search for a LeetCode problem"):
+    search_query_answer = discord.ui.TextInput(
+        label="Enter problem name, ID, or URL:",
+        style=discord.TextStyle.short,
+        required=True,
+    )
+
+    def __init__(self, bot: "DiscordBot") -> None:
+        self.bot = bot
+        super().__init__()
+
+    async def on_submit(self, interaction: discord.Interaction) -> None:
+        await interaction.response.defer()
+
+        embed = await neetcode_search_embed(self.bot, self.search_query_answer.value)
+
+        await interaction.followup.send(embed=embed)
```
Rather than creating a new NeetCodeSearchModal, similar to ProblemSearchModal, let's just make ProblemSearchModal more general. We can do this by add a new param to the constructor of ProblemSearchModal called "submit_embed_creator" or something along those lines, and store it as an attribute. You can then replace the `search_question_embed` in `on_submit` with the `self.submit_embed_creator`, and then we can just pass the appropriate embed creator when we instantiate ProblemSearchModal.

So this allows us to pass the embed creator we want (e.g. search_question_embed with neetcode_search_embed (which I'd rename to search_neetcode_embed), and only have one modal class rather than two similar ones.

Kevin-Roman commented on cogs/neetcode.py:
```diff
@@ -0,0 +1,44 @@
+    @app_commands.command(name="neetcode")
+    async def neetcode(self, interaction: discord.Interaction, language: LanguageField = LanguageField.Python) -> None:
+        """
+        Get neetcode solution of leetcode propblem
```
*problem

Kevin-Roman commented on ui/embeds/neetcode.py:
```diff
@@ -0,0 +1,45 @@
+# Neetcode Search by id
+async def search_neetcode_embed(bot: "DiscordBot", search_text: str, language: Language) -> discord.Embed:
+
+    try:
+        question_title, question_id = await search_question(bot, search_text)
+    except Exception as e:
+        return await question_error_embed()
```
No need to catch exceptions, as ValueErrors and request related exceptions are caught within `search_question` and `bot.http_client.post_data` respectively.

Kevin-Roman replied:
But make sure to check whether either `question_title` and `question_id` are empty, and if so, return nothing.

Kevin-Roman replied:
question_error_embed isn't a coroutine, so no need to await.

Kevin-Roman commented on utils/neetcode.py:
```diff
@@ -0,0 +1,28 @@
+    from bot import DiscordBot
+from constants import Language
+
+async def generate_neetcode_link(bot: "DiscordBot", question_id: str, question_title: str, language: Language):
```
There is no need for this function to be a coroutine, so remove the `async` from this function, and the `await`s used when calling this function.

Kevin-Roman commented on ui/embeds/neetcode.py:
```diff
@@ -0,0 +1,45 @@
+    neetcode_link = await generate_neetcode_link(bot, question_id, question_title, language)
+    response = await bot.http_client.fetch_data(neetcode_link, timeout=10)
+
+    if "Not Found" in response:
+        return await neetcode_error_embed()
```
use:
`if not response_data:`

Kevin-Roman commented on ui/embeds/neetcode.py:
```diff
@@ -0,0 +1,45 @@
+
+    info = await fetch_question_info(bot, question_title)
+    neetcode_link = await generate_neetcode_link(bot, question_id, question_title, language)
+    response = await bot.http_client.fetch_data(neetcode_link, timeout=10)
```
this returns the response data, so rename to `response_data`

Kevin-Roman commented on ui/embeds/neetcode.py:
```diff
@@ -0,0 +1,45 @@
+    return embed
+
+async def neetcode_error_embed() -> discord.Embed:
+    embed = discord.Embed(
+        title="Error",
+        description="Failed to fetch Neetcode solution, or no solution exists",
+        colour=discord.Colour.red(),
+    )
+    return embed
```
Use the `error_embed` from `embeds/common.py`

Kevin-Roman commented on utils/problems.py:
Everything inside this file (except `fetch_daily_question`) are your changes from your other PR, so please remove all the changes back to the original version, and make a commit. In the future, make sure to correctly change branches and stash your changes so that work isn't transferred and accidentally pushed in a commit of another branch.

cf841 commented on cogs/neetcode.py:
```diff
@@ -0,0 +1,44 @@
+        """
+        Get neetcode solution of leetcode propblem
+        """                                                         
+        await interaction.response.send_modal(ProblemSearchModal(self.bot, search_neetcode_embed, language.value))
```
Will do this tomorrow hopefully

Kevin-Roman commented on utils/http_client.py:
```diff
@@ -31,8 +31,9 @@ async def fetch_data(self, *args, **kwargs) -> str | None:
-                if response.status != 404:
+                if response.status == 200:
                     return await response.text()
+            return 
```
no need to return anything here

Kevin-Roman replied:
In order to keep the code consistent, could you match the code used in post_data please. Aka, use `match case` statement, and similarly log f"Fetch request error: (code: {response.status})" if the status code is not 200

Kevin-Roman replied:
In order to keep the code consistent, could you match the code used in post_data please. Aka, use `match case` statement, and similarly log a f"Fetch request error: (code: {response.status})" message if the status code is not 200

cf841 commented:
Cleaned up fetch_data. Only ever seen a 404 in my testing, so should cover all bases

cf841 commented:
- Implemented the function /neetcode, which takes a programming language covered by neetcode, and a problem from leetcode, either by id, name or URL
- Get's the ID and title for the problem, from the leetcode graphql, and uses this to find the corresponding solution in the neetcode git repository. 
- generalized `ProblemSearchModal` so that different commands can call this one function and get the specific data they need from the graphql, by passing an embed function. 
- Added status checking to fetch_data
- Made `search_problem` more useful by returning title and id, rather than just title.

Kevin-Roman commented on ui/embeds/neetcode.py:
```diff
@@ -0,0 +1,45 @@
+        return error_embed()
+
+    embed = discord.Embed(
+        title=f"{question_id}. {info.title} - {language.value.capitalize()} Solution",
```
I think `{question_id}. {info.title} - NeetCode Solution ({language.value.capitalize()})` would better credit NeetCode whilst also mentioning the, not as important, programming language of the solution.

Kevin-Roman reviewed (changes requested):
Request a re-review once you've addressed the feedback provided.

cf841 commented on ui/embeds/neetcode.py:
````diff
@@ -0,0 +1,45 @@
+    embed = discord.Embed(
+        title=f"{question_id}. {info.title} - {language.value.capitalize()} Solution",
+        url=info.link,
+        description=f"```python\n{response_data}```",
````
oops meant to do this and forgot

Kevin-Roman commented on ui/embeds/neetcode.py:
````diff
@@ -0,0 +1,45 @@
+        title=f"{question_id}. {info.title} - {language.value.capitalize()} Solution",
+        url=info.link,
+        description=f"```python\n{response_data}```",
+        colour=discord.Colour.blue(),
````
Make this white

cf841 replied:
no white available, went with next best thing

Kevin-Roman replied:
looks good

Kevin-Roman reviewed (approved):
LGTM!
